llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -705,180 +625,256 @@ void EntryPointGroup::rebuildFromNames(const std::vector<std::string> &Names,
+  // function, which will be used as an entry points group identifier.
+  void registerRule(const std::function<std::string(Function *)> &Callback) {
+    Rules.emplace_back(Rule::RKind::K_Callback, Callback);
+  }
 
-    Ret += "-aspects";
-    for (int A : Aspects) {
-      Ret += "-" + std::to_string(A);
-    }
+  // Creates a simple rule, which adds a value of a string attribute into a
+  // resulting identifier.
+  void registerSimpleStringAttributeRule(StringRef AttrName) {
+    Rules.emplace_back(Rule::RKind::K_SimpleStringAttribute, AttrName);
+  }
 
-    if (UsesLargeGRF)
-      Ret += "-large-grf";
+  // Creates a simple rule, which adds one or another value to a resulting
+  // identifier based on a presence of a metadata on a function.
+  void registerSimpleFlagAttributeRule(StringRef AttrName,
+                                       StringRef IfPresentStr,
+                                       StringRef IfAbsentStr = "") {
+    Rules.emplace_back(Rule::RKind::K_FlagAttribute,
+                       std::tuple{AttrName, IfPresentStr, IfAbsentStr});
+  }
 
-    return Ret;
+  // Creates a simple rule, which adds one or another value to a resulting
+  // identifier based on a presence of a metadata on a function.
+  void registerSimpleFlagMetadataRule(StringRef MetadataName,
+                                      StringRef IfPresentStr,
+                                      StringRef IfAbsentStr = "") {
+    Rules.emplace_back(Rule::RKind::K_FlagMetadata,
+                       std::tuple{MetadataName, IfPresentStr, IfAbsentStr});
   }
 
-  static UsedOptionalFeatures getTombstone() {
-    UsedOptionalFeatures Ret;
-    Ret.IsTombstoneKey = true;
-    return Ret;
+  // Creates a rule, which adds a list of dash-separated integers converted
+  // into strings listed in a metadata to a resulting identifier.
+  void registerListOfIntegersInMetadataRule(StringRef MetadataName) {
+    Rules.emplace_back(Rule::RKind::K_IntegersListMetadata, MetadataName);
   }
 
-  static UsedOptionalFeatures getEmpty() {
-    UsedOptionalFeatures Ret;
-    Ret.IsEmpty = true;
-    return Ret;
+  // Creates a rule, which adds a list of sorted dash-separated integers
+  // converted into strings listed in a metadata to a resulting identifier.
+  void registerListOfIntegersInMetadataSortedRule(StringRef MetadataName) {
+    Rules.emplace_back(Rule::RKind::K_SortedIntegersListMetadata, MetadataName);
   }
 
 private:
-  // For DenseMap:
-  llvm::hash_code Hash = {};
-  bool IsTombstoneKey = false;
-  bool IsEmpty = false;
-
-public:
-  bool operator==(const UsedOptionalFeatures &Other) const {
-    // Tombstone does not compare equal to any other item
-    if (IsTombstoneKey || Other.IsTombstoneKey)
-      return false;
+  struct Rule {
+  private:
+    std::variant<StringRef, std::tuple<StringRef, StringRef, StringRef>,
+                 std::function<std::string(Function *)>>
+        Storage;
+
+  public:
+    enum class RKind {
```

Comment:
Please add a comment on what these kinds stand for. Thanks

Reply:
Added brief descriptions in 13169d46b86d40196060de52bfe1dff86d75409f